services/masterdata/src/lib/components/masterdatas-edit-form/masterdatas-edit-form.component.ts

```diff
@@ -10,149 +11,157 @@ import { MasterdataMaxLength } from "../../core/utils";
+		this.needsTypeIdSelector = !(this.typeId !== undefined);
+
+		if (this.typeId) this.masterdataType$ = this.service.getMasterdataTypeById(this.typeId);
+
+		if (!this.isAddForm) {
+			this.loadData();
+			this.saveBtnText = "Update";
+			this.formTitle = "Edit";
+		}
+		if (this.isAddForm && this.needsTypeIdSelector) {
+			this.loadTypesList();
+		}
+
+		const whenAddForm = this.isAddForm && {
+			masterdataTypeId: [this.typeId ?? "", [Validators.required]],
+			key: ["", [Validators.required, Validators.maxLength(this.maxLength.key)]]
+		};
+		this.dataForm = this.formBuilder.group({
+			...whenAddForm,
+			value: ["", [Validators.required, Validators.maxLength(this.maxLength.value)]],
+			name: ["", [Validators.required, Validators.maxLength(this.maxLength.name)]],
+			description: ["", [Validators.maxLength(this.maxLength.description)]],
+			metadata: ["", [Validators.maxLength(this.maxLength.metadata)]]
+		});
+	}
+
+	makeMetadataEditorOptions(): JsonEditorOptions {
+		const opt = new JsonEditorOptions();
+		opt.mode = "tree";
+		opt.modes = ["code", "text", "tree", "view"];
+		return opt;
+	}
+
+	// convenience getter for easy access to form fields
+	get getFormFields() {
+		return this.dataForm.controls;
+	}
+
+	loadData() {
+		this.isLoading = true;
+		this.service.getMasterdataById(this.typeId, this.id).subscribe((data) => {
+			this.dataForm.patchValue({
+				value: data.value,
+				name: data.name,
+				description: data.description,
+				metadata: data.metadata
+			});
+			this.item = data || {};
+			this.isLoading = false;
+		});
+	}
+
+	onSubmit() {
+		this.isFormSubmitted = true;
+		if (this.dataForm.invalid) {
+			// stop here if form is invalid
+			return;
+		}
+
+		this.isLoading = true;
+		if (this.isAddForm) {
+			const key = getRequiredFieldValue<string>(this.dataForm, "key");
+			const masterdataTypeIdSelector = getRequiredFieldValue<KeyValuePair<string, string>>(this.dataForm, "masterdataTypeId").key;
+			const masterdataTypeId = getRequiredFieldValue<string>(this.dataForm, "masterdataTypeId");
+			const value = getRequiredFieldValue<string>(this.dataForm, "value");
+			const name = getRequiredFieldValue<string>(this.dataForm, "name");
+			const description = getFieldValue<string>(this.dataForm, "description");
+			const metadata = getFieldValue<string>(this.dataForm, "metadata");
+
+			const model = {} as IMasterdataCreate;
+			model.key = key;
+			model.masterdataTypeId = masterdataTypeIdSelector ?? masterdataTypeId ?? undefined;
+			model.value = value;
+			model.name = name;
+			model.description = description;
+			model.metadata = metadata;
+
+			// const model = this.dataForm.value as IMasterdataCreate;
+			this.service.createMasterdata(model).subscribe(() => {
+				this.btnCancel();
+				this.isLoading = false;
+				this.toastService.success("Add masterdata", "The masterdata was succesfully added.");
+			});
+		} else {
+			const value = getRequiredFieldValue<string>(this.dataForm, "value");
```

Comment:
@dageorgero  Remove duplicate code by `if`-ing the changes.

Reply:
removed some of the duplicated code!